Add template files with ha-sap-terraform-deployments project

nick-wang opened the pull request:
1. Polish the existed template and spec file.

2. Add the DRBD/NFS templates with ha-sap-terraform-deployments.
    Support libvirt/gcp/azure/aws provider.
    Leave the [cluster examples](https://github.com/SUSE/drbd-formula/tree/master/examples/with_pacemaker) since DRBD HA cluster may run without ha-sap-terraform-deployments
    
    Fix https://github.com/SUSE/drbd-formula/issues/15

nick-wang committed:
Polish the drbd template format

nick-wang committed:
Add template files with ha-sap-terraform-deployments project

Add the DRBD/NFS templates with ha-sap-terraform-deployments.
Support libvirt/gcp/azure/aws provider.

Fix https://github.com/SUSE/drbd-formula/issues/15

nick-wang committed:
Clean the spec and bump up version to 0.4.2

nick-wang commented:
The new pillar files/template are come from https://github.com/SUSE/ha-sap-terraform-deployments/releases/tag/7.0.0
Plan to remove the template(leave the pillar files in automatic) from ha-sap-terraform-deployments after merged.

arbulu89 commented on drbd-formula.spec:
```diff
@@ -60,9 +57,9 @@ cp -R examples %{buildroot}%{fdir}/states/%{fname}/%{ftemplates}/
-%defattr(-,root,root,-)
 %if 0%{?sle_version} < 120300
-%doc README.md LICENSE
+%license LICENSE
```
The reason why we added this `if` statement is that before SLE12SP3 the `license` tag doesn't exist.
Actually, as we don't build anymore for versions less than SLE12SP3, I would remove this. We would have:
```
%doc README.md
%license LICENSE

```

nick-wang replied:
I made a mistake. Follow the spec cleaner without check too much, my bad to make the regression.

arbulu89 commented on examples/with_ha-sap-terraform-deployments/pillar.example.cluster_cloud:
```diff
@@ -0,0 +1,64 @@
+cluster:
```
I wouldn't personally add this file.
It has a lot of references to the `ha-sap-terraform-deployments` project. Actually, it would only work in that project. I would just keep this in that project

arbulu89 reviewed (changes requested):
@nick-wang The idea is good. I have some things that I would prefer to change.

- I wouldn't add the `pillar.example.cluster_cloud` file. It doesn't belong to this project in my opinion
- I would rename some templates:
  1. `examples/with_ha-sap-terraform-deployments/drbd_with_nfs_cloud.j2` -> `examples/cluster_resources_nfs_cloud.j2`
  2. `examples/with_ha-sap-terraform-deployments/pillar.example.drbd_cloud` -> `pillar.example`
  3. `examples/with_pacemaker/drbd_with_nfs_crm.js` -> examples/cluster_resources_nfs.j2`

I guess that the cloud and non-cloud templates are created for simplicity, right? So if we use libvirt for example, we would need to use the non cloud version

nick-wang committed:
Rename the templates and folder structure.

arbulu89 commented on drbd-formula.spec:
```diff
@@ -60,7 +57,6 @@ cp -R examples %{buildroot}%{fdir}/states/%{fname}/%{ftemplates}/
 
 %files
-%defattr(-,root,root,-)
 %if 0%{?sle_version} < 120300
```
As commented, I would take advantage of these fixes and remove the `< 120300` condition. We don't build anymore for versions less than SLE12SP3, and maintenance has been a bit picky in these things.
The next would be just fine:
```
%doc README.md
%license LICENSE
```
But well, we can do this any other time

nick-wang replied:
Got it, i will fix it this time.

arbulu89 reviewed (changes requested):
Hi @nick-wang ,

I still don't like the folder naming. I wouldn't put references to `ha-sap-terraform-deployments`. This makes the user think that this templates are only for that project, and that's not true.
I would just simply remove the `with_ha-sap-terraform-deployments` and `with_pacemaker` folders. I would put all the templates directly in the `templates` folder.
In fact, I would rename `examples` by `templates`

nick-wang commented:
> I still don't like the folder naming. I wouldn't put references to `ha-sap-terraform-deployments`. This makes the user think that this templates are only for that project, and that's not true.
> I would just simply remove the `with_ha-sap-terraform-deployments` and `with_pacemaker` folders. I would put all the templates directly in the `templates` folder.
> In fact, I would rename `examples` by `templates`

@arbulu89 The reason why i split `with_ha-sap-terraform-deployments` folder is because the template `cluster_resources_nfs_cloud.j2` and corresponding pillar file has too many elements of `cloud provider`, which may confused user and add complexity when customized the configuration. Though it is actually have condition to handler the case that `cloud_provider` not in `[aws, gcp, azure]`.

I rename the `with_ha-sap-terraform-deployments` to `with_pacemaker_cloud`. If you still think it is not good to keep the template(https://github.com/nick-wang/drbd-formula/blob/issue-15/examples/with_pacemaker/cluster_resources_nfs.j2). I can merge the folders of `with pacemaker`. Have one with `nfs formula` and the other one with `pacemaker habootstrap formula`

Will rename `examples` to `templates`

arbulu89 reviewed (approved):
@nick-wang 
I don't still like the naming, but we can go with it no issues.
If you don't want to put them in the same folder, let's at least rename too:
- `with_ha-sap-terraform-deployments/cluster_resources_nfs_cloud.j2` -> `habootstrap-formula/cluster_resources_nfs_cloud.j2`
- `with_ha-sap-terraform-deployments/notify-split-brain-haclusterexporter-suse-metric.sh` -> `ha_cluster_exporter/notify-split-brain-haclusterexporter-suse-metric.sh`
- (Both in `templates` folder). I don't want the `ha-sap-terraform-deployments` reference, because this is just a casuality, the templates are to be used by the `habootstrap-formula` and the `ha_cluster_exporter`.
**PD: I see more and more a lot of confusion having the cluster resources in the formulas that are not the cluster formula. I'm considering to move all the cluster resource templates there. In the past, they were put in these folders for some business reason, but I think they don't make sense. We will need to discuss this**

nick-wang commented:
> @nick-wang
> I don't still like the naming, but we can go with it no issues.
> If you don't want to put them in the same folder, let's at least rename too:
> 
> * `with_ha-sap-terraform-deployments/cluster_resources_nfs_cloud.j2` -> `habootstrap-formula/cluster_resources_nfs_cloud.j2`
> * `with_ha-sap-terraform-deployments/notify-split-brain-haclusterexporter-suse-metric.sh` -> `ha_cluster_exporter/notify-split-brain-haclusterexporter-suse-metric.sh`
> * (Both in `templates` folder). I don't want the `ha-sap-terraform-deployments` reference, because this is just a casuality, the templates are to be used by the `habootstrap-formula` and the `ha_cluster_exporter`.
>   **PD: I see more and more a lot of confusion having the cluster resources in the formulas that are not the cluster formula. I'm considering to move all the cluster resource templates there. In the past, they were put in these folders for some business reason, but I think they don't make sense. We will need to discuss this**

@arbulu89 Thanks a lot for the idea.
I do agree with `habootstrap-formula` sounds better.

I changed the folder structure again, now is like:
```
templates # tree
.
├── habootstrap-formula
│   ├── cluster_resources_nfs_cloud.j2
│   ├── cluster_resources_nfs.j2
│   └── pillar.example.cluster
├── ha_cluster_exporter
│   └── notify-split-brain-haclusterexporter-suse-metric.sh
├── nfs-formula
│   └── pillar.example.nfs
├── pillar.example.drbd
└── README.md
```
How do you think of it?

arbulu89 reviewed (approved):
Great! Thanks @nick-wang

nick-wang commented:
@arbulu89 Thanks for the help. I will do some test(build/install) before click merge. (However, i think it should be fine)

arbulu89 commented:
> @arbulu89 Thanks for the help. I will do some test(build/install) before click merge. (However, i think it should be fine)

Nice! Keep in mind that afterwards you will need to update the `ha-sap-terraform-deployments` project to use this new structure and remove duplicated files XD

nick-wang commented:
> > @arbulu89 Thanks for the help. I will do some test(build/install) before click merge. (However, i think it should be fine)
> 
> Nice! Keep in mind that afterwards you will need to update the `ha-sap-terraform-deployments` project to use this new structure and remove duplicated files XD

@arbulu89 Yes, i was plan to test this with changed `ha-sap-terraform-deployments`. Only merge this after all finished.

Will adapt(remove) the j2 template and exporter script with the one in `drbd-formula`. Only leave the pillar file in `automatic` for pre_deployment. The two change should be used together with v8. I think besides the release note of `ha-sap-terraform-deployments`, also better to maintain a `requirements.txt` to list the dependencies. 
How do you think?

nick-wang committed:
Remove the duplicate template.

/usr/share/salt-formulas/states/drbd/templates/templates to
/usr/share/salt-formulas/states/drbd/templates/

nick-wang commented:
Corresponding change in `ha-sap-terraform-deployments` project is https://github.com/SUSE/ha-sap-terraform-deployments/pull/663

nick-wang commented:
From this pull request, any change of the path of template need to align with project `ha-sap-terraform-deployments`.
Refer to https://github.com/SUSE/ha-sap-terraform-deployments/pull/663
